# Markdown 1 

*Here we can observe that the DataSet provided indeed has missing values. The code has read each individual column and the column 'id' is missing 1 value and the column 'song' is missing 11 values*

In [19]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

# Markdown 2 
*With the help of CoPilot, we can define the words "observations" and "variables" within the context of the dataset. 

*The term "observations" are defined to refer to the individual rows of the dataset. In the context of this data set for example, the rows grant information/observations about each individual villager. 

*The term "variables" are defined to refer to the columns of the dataset. For example, the columns in the data set grant information about variables regarding all the villagers (personality, birthday, species, etc).


*The code below (with the help of CoPilot) has printed the number of rows and columns of the dataset. 

In [20]:
import pandas as pd

import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
data = pd.read_csv(url)

# Display the column names and the number of rows in the dataset
columns = data.columns.tolist()
num_rows = len(data)

print("Columns:", columns)
print("Number of rows:", num_rows)

Columns: ['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality', 'song', 'phrase', 'full_id', 'url']
Number of rows: 391


# Markdown 3

In [21]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
data = pd.read_csv(url)

species_counts = data['name'].value_counts()
print(species_counts)

name
Admiral    1
Muffy      1
Paula      1
Patty      1
Pate       1
          ..
Elvis      1
Eloise     1
Elmer      1
Ellie      1
Zucker     1
Name: count, Length: 391, dtype: int64


In [22]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
data = pd.read_csv(url)

species_counts = data['species'].value_counts()
print(species_counts)

species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64


In [23]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
data = pd.read_csv(url)

species_counts = data['gender'].value_counts()
print(species_counts)

gender
male      204
female    187
Name: count, dtype: int64


In [24]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
data = pd.read_csv(url)

species_counts = data['birthday'].value_counts()
print(species_counts)

birthday
1-27     2
12-5     2
7-31     2
3-26     2
8-3      2
        ..
4-3      1
10-26    1
7-23     1
12-8     1
3-8      1
Name: count, Length: 361, dtype: int64


In [14]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
data = pd.read_csv(url)

species_counts = data['id'].value_counts()
print(species_counts)

id
admiral    1
mott       1
paula      1
patty      1
pate       1
          ..
eloise     1
elmer      1
ellie      1
elise      1
zucker     1
Name: count, Length: 390, dtype: int64


In [15]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
data = pd.read_csv(url)

species_counts = data['personality'].value_counts()
print(species_counts)

personality
lazy      60
normal    59
cranky    55
snooty    55
jock      55
peppy     49
smug      34
uchi      24
Name: count, dtype: int64


In [16]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
data = pd.read_csv(url)

species_counts = data['song'].value_counts()
print(species_counts)

song
K.K. Country     10
Forest Life       9
Imperial K.K.     7
K.K. Soul         7
K.K. Ragtime      7
                 ..
Aloha K.K.        2
Drivin'           1
Senor K.K.        1
K.K.  Bazaar      1
K.K. D&B          1
Name: count, Length: 92, dtype: int64


In [31]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
data = pd.read_csv(url)

species_counts = data['phrase'].value_counts()
print(species_counts)

phrase
wee one       2
quacko        2
bloop         2
aye aye       1
snoot         1
             ..
lambchop      1
yeah buddy    1
chow down     1
unh-hunh      1
pronk         1
Name: count, Length: 388, dtype: int64


In [32]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
data = pd.read_csv(url)

species_counts = data['full_id'].value_counts()
print(species_counts)

full_id
villager-admiral    1
villager-muffy      1
villager-paula      1
villager-patty      1
villager-pate       1
                   ..
villager-elvis      1
villager-eloise     1
villager-elmer      1
villager-ellie      1
villager-zucker     1
Name: count, Length: 391, dtype: int64


# Markdown 4

In [27]:
import pandas as pd 
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)


rows, columns = df.shape
print(f"Number of rows: {rows}, Number of columns: {columns}")

Number of rows: 891, Number of columns: 15


In [28]:
import pandas as pd 
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)


all_summary = df.describe(include='all')
print(all_summary)

          survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male deck  embark_town alive  \
count   8

*Here, we can analyze that df.shape and df.describe() give a different analysis of the dataset presented in regards to the number of columns and the value count reported under these columns.

*df.shape simply analyzes the data and reports that there are 15 columns and 891 rows. In theory, this should imply that each of the 15 columns in the data set has 891 recorded values.

*However we do not see this in df.describe() regarding the count for each column. The age, embarked, deck, and embark_town columns all have count values below 891. This confirms that these specific columns have missing values.  

# Markdown 5

*The difference between an attribute and a method are significant in the sense that both perform vastly different tasks with regards to a dataset. 

*An "attribute" such as df.shape simply states how many columns and rows there are in a dataset. However, it does not do any calculations and no further information is given in regards to the dataset. It simply tells us how the dataset is structured.

*A "method" such as df.describe() performs specific calculations on a dataset and gives more direct statistical information compared to an attribute. The calculation operations one can perform vary (see Markdown 6), thus the () is included to specify which calculation one wants to compute. 

# Markdown 6

*Count refers to the # of valid non-missing values in the dataset

*Mean refers to the average of the values in the dataset

*std (standard deviation) refers to the dispersion/spread of the data. A high std means the values in the dataset are spread out, while a low std means the values in the dataset are closer to the average

*min (minimum) refers to the smallest value in the dataset 

*max (maximum) refers to the largest value in the dataset

*25% (1st quartile) refers to the number where 1/4 of the smallest values in the data fall below this point

*50% (2nd quartile) refers to the number holding the middle point of the data. The data values are split 50/50 below/above this point.

*75% (3rd quartile) refers to the number where 3/4 of the values fall below this point. The largest 25% of values in the dataset are above this point.

# CoPilot Summary (Questions 1-6)

*A quoted summary from CoPilot regarding our session together with this assignment

"Sure! Here’s a summary of our interactions:

Dataset Overview:
You downloaded a dataset about characters from Animal Crossing and wanted to know the columns and the amount of data.
We attempted to load the dataset from the provided URL but encountered an issue. I provided a Python code snippet to load the dataset locally and display the column names and number of rows.

Understanding Observations and Variables:
We discussed the meanings of “observations” (rows) and “variables” (columns) in the context of pandas DataFrames.
I provided a hypothetical table to visualize these concepts.

Using df.describe() and df['column'].value_counts():
We explored how to use df.describe() for numeric summaries and df['column'].value_counts() for categorical data.
I explained that df.describe() does not require any arguments and provided a code example.

Handling Large Datasets:
We discussed analyzing each column individually for large datasets.
I provided a Python function to summarize each column, either using describe() for numeric columns or value_counts() for categorical columns.

Discrepancies Between df.shape and df.describe():
We explored the differences between the dataset size given by df.shape and what is reported by df.describe(), especially with non-numeric variables and missing values.
I explained how df.describe() only includes numeric columns by default and how the “count” in df.describe() reflects non-missing values.

Attributes vs. Methods:
We discussed the difference between an “attribute” (e.g., df.shape) and a “method” (e.g., df.describe()).
I provided definitions and examples to clarify the difference.

Types of Calculations with Methods:
We reviewed various calculations and operations that can be performed with pandas methods, including descriptive statistics, aggregation, data manipulation, filtering, group by operations, handling missing data, and merging/joining DataFrames.

If you need any more details or further clarification on any of these points, feel free to ask!"

# Markdown 7 

*1. A use case in which we would use 'df.dropna()' over 'del df['col']' would be in a scenario where a state keeps information on voter registration.

*Let's say that out of 100 voters in a county, one of the voters is missing their (numerical) legal age under the "age" column. Using 'del df['col']' in this case would be illogical and irresponsible since the 'age' column would be deleted. Not only would this delete the column, but it would also wipe out crucial information on all other registered voters.

*With 'df.dropna()' we would simply purge the voter with irregular voter information from the voter registration data. Not only would this correctly remove a row of information on an individual with incorrect data to vote (since they shouldn't be voting), but it would also not interfere with the other 99 individuals who have correct voter registration information. 